boot/boot.go

```diff
@@ -0,0 +1,189 @@
+// returns it. The second return value indicates whether no boot participant
+// exists, and if false the first return value will be nil.
+//
+// Currently, on classic, nothing is a boot participant (applicable
+// will always be false).
+func Lookup(s snap.PlaceInfo, t snap.Type, model Model, onClassic bool) (bp BootParticipant, applicable bool) {
+	if onClassic {
+		return nil, false
+	}
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase {
+		return nil, false
+	}
+
+	if model != nil {
+		switch t {
+		case snap.TypeKernel:
+			if s.InstanceName() != model.Kernel() {
+				return nil, false
+			}
+		case snap.TypeBase, snap.TypeOS:
+			base := model.Base()
+			if base == "" {
+				base = "core"
+			}
+			if s.InstanceName() != base {
+				return nil, false
+			}
+		}
+	}
+
+	cbp := &coreBootParticipant{s: s, t: t}
+	bp = cbp
+	if t == snap.TypeKernel {
+		bp = &coreKernel{cbp}
+	}
+
+	return bp, true
+}
+
+// InUse checks if the given name/revision is used in the
+// boot environment
+func InUse(name string, rev snap.Revision) bool {
+	bootloader, err := bootloader.Find()
+	if err != nil {
+		logger.Noticef("cannot get boot settings: %s", err)
+		return false
+	}
+
+	bootVars, err := bootloader.GetBootVars("snap_kernel", "snap_try_kernel", "snap_core", "snap_try_core")
+	if err != nil {
+		logger.Noticef("cannot get boot vars: %s", err)
+		return false
+	}
+
+	snapFile := filepath.Base(snap.MountFile(name, rev))
+	for _, bootVar := range bootVars {
+		if bootVar == snapFile {
+			return true
+		}
+	}
+
+	return false
+}
+
+var (
+	ErrBootNameAndRevisionAgain = errors.New("boot revision not yet established")
+)
+
+type NameAndRevision struct {
+	Name     string
+	Revision snap.Revision
+}
+
+// GetCurrentBoot returns the currently set name and revision for boot for the given
+// type of snap, which can be snap.TypeBase (or snap.TypeOS), or snap.TypeKernel.
+// Returns ErrBootNameAndRevisionAgain if the values are temporarily not established.
+func GetCurrentBoot(t snap.Type) (*NameAndRevision, error) {
+	var bootVar, errName string
+	switch t {
+	case snap.TypeKernel:
+		bootVar = "snap_kernel"
+		errName = "kernel"
+	case snap.TypeOS, snap.TypeBase:
+		bootVar = "snap_core"
+		errName = "base"
+	default:
+		return nil, fmt.Errorf("internal error: cannot find boot revision for snap type %q", t)
+	}
+
+	loader, err := bootloader.Find()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get boot settings: %s", err)
+	}
+
+	m, err := loader.GetBootVars(bootVar, "snap_mode")
+	if err != nil {
+		return nil, fmt.Errorf("cannot get boot variables: %s", err)
+	}
+
+	if m["snap_mode"] == "trying" {
+		return nil, ErrBootNameAndRevisionAgain
+	}
+
+	nameAndRevno, err := nameAndRevnoFromSnap(m[bootVar])
+	if err != nil {
+		return nil, fmt.Errorf("cannot get name and revision of boot %s: %v", errName, err)
+	}
+
+	return nameAndRevno, nil
+}
+
+func nameAndRevnoFromSnap(sn string) (*NameAndRevision, error) {
```

Comment:
I would add a comment that documents the input format, just for simplicity of scanning that vs decoding the function body.